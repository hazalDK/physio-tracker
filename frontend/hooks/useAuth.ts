import * as SecureStore from "expo-secure-store";
import axios from "axios";
import { useNavigation } from "@react-navigation/native";
import { StackNavigationProp } from "@react-navigation/stack";
import { RootStackParamsList } from "@/types/navigation";
import { Alert } from "react-native";

export function useAuth() {
  const navigation =
    useNavigation<StackNavigationProp<RootStackParamsList, "login">>();

  const getToken = async () => {
    return await SecureStore.getItemAsync("access_token");
  };

  const refreshToken = async () => {
    try {
      const refreshToken = await SecureStore.getItemAsync("refresh_token");
      if (!refreshToken) {
        throw new Error("No refresh token available");
      }

      const refreshUrl = process.env.API_URL || "http://192.168.68.111:8000";
      const refreshResponse = await axios.post(
        `${refreshUrl}/api/token/refresh/`,
        { refresh: refreshToken }
      );

      const newToken = refreshResponse.data.access;
      const newRefreshToken = refreshResponse.data.refresh;

      await Promise.all([
        SecureStore.setItemAsync("access_token", newToken),
        SecureStore.setItemAsync("refresh_token", newRefreshToken),
      ]);

      return newToken;
    } catch (error) {
      await Promise.all([
        SecureStore.deleteItemAsync("access_token"),
        SecureStore.deleteItemAsync("refresh_token"),
      ]);

      Alert.alert("Session Expired", "Please login again", [
        { text: "OK", onPress: () => navigation.navigate("login") },
      ]);

      return null;
    }
  };

  const createApiInstance = async () => {
    let token = await getToken();

    if (!token) {
      Alert.alert("Login Required", "Please sign in to continue", [
        { text: "OK", onPress: () => navigation.navigate("login") },
      ]);
      return null;
    }

    return axios.create({
      baseURL: process.env.API_URL || "http://192.168.68.111:8000",
      timeout: 10000,
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
    });
  };

  return { getToken, refreshToken, createApiInstance };
}
