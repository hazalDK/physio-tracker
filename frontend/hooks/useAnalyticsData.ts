import { useState, useEffect } from "react";
import axios from "axios";
import { useAuth } from "./useAuth";

// Helper function to format date
const formatDate = (date: Date) => {
  return date.toISOString().split("T")[0];
};

// Type for analytics data (can be either adherence or pain data)
interface ChartData {
  labels: string[];
  datasets: {
    data: number[];
  }[];
}

/**
 * Custom hook to fetch and manage analytics data (adherence or pain levels).
 * @param {string} endpoint - The API endpoint to fetch data from.
 * @param {"adherence" | "pain"} dataType - The type of data to fetch (either adherence or pain).
 * @returns {Object} - An object containing loading state, chart data, average, history, error, and functions to navigate weeks.
 * @throws {Error} - Throws an error if the API call fails.
 * */
export function useAnalyticsData<T>(
  endpoint: string,
  dataType: "adherence" | "pain"
): object {
  const [loading, setLoading] = useState(true);
  const [chartData, setChartData] = useState<ChartData>({
    labels: ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"],
    datasets: [{ data: [0, 0, 0, 0, 0, 0, 0] }],
  });
  const [average, setAverage] = useState(0);
  const [history, setHistory] = useState<T[]>([]);
  const [error, setError] = useState<string | null>(null);

  // Add week navigation state
  const [currentEndDate, setCurrentEndDate] = useState(new Date());
  const [weekTitle, setWeekTitle] = useState("Current Week");
  const { createApiInstance, refreshToken } = useAuth();

  // Function to go to previous week
  const goToPreviousWeek = () => {
    const newEndDate = new Date(currentEndDate);
    newEndDate.setDate(newEndDate.getDate() - 7);
    setCurrentEndDate(newEndDate);
    fetchData(newEndDate);
    updateWeekTitle(newEndDate);
  };

  // Function to go to next week (but not beyond current date)
  const goToNextWeek = () => {
    const today = new Date();
    const newEndDate = new Date(currentEndDate);
    newEndDate.setDate(newEndDate.getDate() + 7);

    // Don't go beyond current date
    if (newEndDate > today) {
      newEndDate.setTime(today.getTime());
    }

    setCurrentEndDate(newEndDate);
    fetchData(newEndDate);
    updateWeekTitle(newEndDate);
  };

  // Update week title based on date range
  const updateWeekTitle = (endDate: Date) => {
    const startDate = new Date(endDate);
    startDate.setDate(startDate.getDate() - 6);

    const today = new Date();
    const isCurrentWeek =
      today.getDate() === endDate.getDate() &&
      today.getMonth() === endDate.getMonth() &&
      today.getFullYear() === endDate.getFullYear();

    if (isCurrentWeek) {
      setWeekTitle("Current Week");
    } else {
      const startMonth = startDate.toLocaleString("default", {
        month: "short",
      });
      const endMonth = endDate.toLocaleString("default", { month: "short" });

      if (startMonth === endMonth) {
        setWeekTitle(
          `${startMonth} ${startDate.getDate()} - ${endDate.getDate()}`
        );
      } else {
        setWeekTitle(
          `${startMonth} ${startDate.getDate()} - ${endMonth} ${endDate.getDate()}`
        );
      }
    }
  };

  // Function to fetch data for a specific end date
  const fetchData = async (endDate: Date) => {
    setLoading(true);
    try {
      const api = await createApiInstance();
      if (!api) return;

      try {
        // Fetch stats from the API with date parameter
        const response = await api.get(
          `${endpoint}?end_date=${formatDate(endDate)}`
        );
        if (response.data) {
          setChartData(response.data.chart_data);

          if (dataType === "adherence") {
            setAverage(response.data.average_adherence);
          } else {
            setAverage(response.data.average_pain);
          }

          setHistory(response.data.history);
        }
      } catch (error) {
        // Handle 401 error by refreshing the token and retrying the request
        if (axios.isAxiosError(error) && error.response?.status === 401) {
          // Attempt token refresh
          const newToken = await refreshToken();
          if (!newToken) return;

          const api = await createApiInstance();
          if (!api) return;

          // Retry with new token
          const response = await api.get(
            `${endpoint}?end_date=${formatDate(endDate)}`
          );

          if (response.data) {
            setChartData(response.data.chart_data);

            if (dataType === "adherence") {
              setAverage(response.data.average_adherence);
            } else {
              setAverage(response.data.average_pain);
            }

            setHistory(response.data.history);
          }
        } else {
          // Handle other errors
          console.error("API request failed:", error);
          setError(
            `Failed to load your ${
              dataType === "adherence" ? "exercises" : "pain data"
            }. Please check your connection and try again.`
          );
        }
      }
    } catch (error) {
      console.error(
        `Failed to load your ${
          dataType === "adherence" ? "exercises" : "pain data"
        }.`
      );
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    // Initial data fetch for current week
    fetchData(currentEndDate);
    updateWeekTitle(currentEndDate);
  }, []);

  return {
    loading,
    chartData,
    average,
    history,
    error,
    weekTitle,
    goToPreviousWeek,
    goToNextWeek,
  };
}
