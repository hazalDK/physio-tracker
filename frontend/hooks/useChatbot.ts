import { useState, useEffect, useCallback } from "react";
import { Alert } from "react-native";
import Toast from "react-native-root-toast";
import { messageType } from "@/types/chatbot";
import { createChatApi, apiRequestWithRefresh } from "../api";
import { useAuth } from "./useAuth";

export default function useChatbot() {
  const auth = useAuth();
  const [messages, setMessages] = useState<messageType[]>([]);
  const [userExercises, setUserExercises] = useState([]);
  const [recentPain, setRecentPain] = useState(0);
  const [isLoading, setIsLoading] = useState(false);
  const [isInitialLoading, setIsInitialLoading] = useState(true);
  const [api, setApi] = useState<any>(null);

  // Initialize API and chat
  useEffect(() => {
    const initializeApi = async () => {
      const apiInstance = await auth.createApiInstance();
      setApi(apiInstance);
    };

    initializeApi();

    // Set initial messages
    setMessages([
      {
        id: "disclaimer",
        text: "DISCLAIMER",
        sender: "disclaimer",
      },
      {
        id: "welcome",
        text: "Welcome to your physiotherapy assistant! How can I help you today?",
        sender: "bot",
      },
    ]);
  }, []);

  // Fetch user data when API is available
  useEffect(() => {
    if (api) {
      fetchUserData();
    }
  }, [api]);

  const fetchUserData = async () => {
    if (!api) return;

    setIsInitialLoading(true);
    const chatApi = createChatApi(api);

    try {
      // Use the wrapper function for API calls with refresh
      const fetchData = async () => {
        const [reportsData, exercisesData] = await Promise.all([
          chatApi.fetchReports(),
          chatApi.fetchUserExercises(),
        ]);

        return { reportsData, exercisesData };
      };

      const { reportsData, exercisesData } = await apiRequestWithRefresh(
        fetchData,
        auth
      );

      setUserExercises(exercisesData);

      if (reportsData.length > 0) {
        setRecentPain(reportsData[0].pain_level);
      }
    } catch (error) {
      console.error("Failed to fetch user data:", error);

      Alert.alert(
        "Error",
        "Failed to load your exercises. Please check your connection and try again."
      );
    } finally {
      setIsInitialLoading(false);
    }
  };

  const sendMessage = async (inputText: string) => {
    if (!inputText.trim() || !api) return;

    // Add user message to chat
    const userMessageId = Date.now().toString();
    setMessages((prev) => [
      ...prev,
      { id: userMessageId, text: inputText, sender: "user" },
    ]);

    setIsLoading(true);
    const chatApi = createChatApi(api);

    try {
      // Create context object with user's exercise data
      const exerciseContext = {
        exercises: userExercises,
        recentPain: recentPain,
      };

      // Use the wrapper function for API calls with refresh
      const sendMessageCall = async () => {
        return await chatApi.sendChatMessage(inputText, exerciseContext);
      };

      // Send message to API with refresh handling
      const response = await apiRequestWithRefresh(sendMessageCall, auth);

      // Add bot response to chat
      setMessages((prev) => [
        ...prev,
        {
          id: Date.now().toString(),
          text: response.message,
          sender: "bot",
        },
      ]);
    } catch (error) {
      console.error("Error sending message:", error);

      setMessages((prev) => [
        ...prev,
        {
          id: Date.now().toString(),
          text: "Sorry, I encountered an error. Please try again.",
          sender: "bot",
        },
      ]);
    } finally {
      setIsLoading(false);
    }
  };

  const resetChat = async () => {
    if (!api) return;

    const chatApi = createChatApi(api);

    try {
      // Use the wrapper function for API calls with refresh
      const resetCall = async () => {
        await chatApi.resetChat();
      };

      await apiRequestWithRefresh(resetCall, auth);

      // Reset messages
      setMessages([
        {
          id: "disclaimer",
          text: "DISCLAIMER",
          sender: "disclaimer",
        },
        {
          id: "welcome",
          text: "Welcome to your physiotherapy assistant! How can I help you today?",
          sender: "bot",
        },
      ]);

      // Refresh user data
      fetchUserData();

      Toast.show("Chat reset!", {
        duration: Toast.durations.SHORT,
        position: Toast.positions.BOTTOM,
        backgroundColor: "#14b8a6",
        textColor: "white",
        shadow: true,
      });
    } catch (error) {
      console.error("Failed to reset chat:", error);
      Alert.alert("Error", "Could not reset chat.");
    }
  };

  return {
    messages,
    isLoading,
    isInitialLoading,
    sendMessage,
    resetChat,
  };
}
