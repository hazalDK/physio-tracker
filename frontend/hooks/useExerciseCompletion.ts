import { useState } from "react";
import { Alert } from "react-native";
import { useNavigation } from "@react-navigation/native";
import { useAuth } from "./useAuth";

// Custom hook to manage user profile data and related operations.
export function useExerciseCompletion(userExerciseId: number | null) {
  const [showCompletionForm, setShowCompletionForm] = useState(false);
  const [reps, setReps] = useState(0);
  const [sets, setSets] = useState(0);
  const [painLevel, setPainLevel] = useState(0);
  const navigation = useNavigation();
  const { createApiInstance } = useAuth();

  // Function to handle saving exercise completion data
  // This function is called when the user submits the completion form.
  const handleSave = async () => {
    // Validate input fields
    if (!reps || !sets || !painLevel) {
      Alert.alert("Error", "Please fill all required fields");
      return;
    }

    // Api instance creation
    const api = await createApiInstance();
    if (!api) return;

    try {
      // Update user exercise data
      await api.put(`/user-exercises/${userExerciseId}/`, {
        reps: reps,
        sets: sets,
        pain_level: painLevel,
        completed: true,
      });

      // Closes the modal
      setShowCompletionForm(false);

      // Handle pain level logic
      // If pain level is low (1-3), check if they can increase difficulty
      if (painLevel <= 3) {
        try {
          const canIncreaseResponse = await api.get(
            `/user-exercises/${userExerciseId}/can_increase/`
          );

          if (canIncreaseResponse.data.can_increase) {
            Alert.alert(
              "Exercise Progress",
              "Your pain level is low and you've had consistent low pain for 3 days. Would you like to increase the difficulty for next time?",
              [
                {
                  text: "No",
                  onPress: () => navigation.goBack(),
                  style: "cancel",
                },
                {
                  text: "Yes",
                  onPress: async () => {
                    try {
                      await api.post(
                        `/user-exercises/${userExerciseId}/confirm_increase/`,
                        { confirm: "yes" }
                      );
                      Alert.alert(
                        "Success",
                        "Exercise difficulty increased for next time!"
                      );
                      navigation.goBack();
                    } catch (error) {
                      console.error("Error increasing difficulty:", error);
                      Alert.alert(
                        "Error",
                        "Failed to increase difficulty. Please try again."
                      );
                      navigation.goBack();
                    }
                  },
                },
              ]
            );
          } else {
            Alert.alert(
              "Exercise Progress",
              "Your pain level is low. Keep it up for 3 days to unlock higher difficulty!",
              [{ text: "OK", onPress: () => navigation.goBack() }]
            );
          }
        } catch (error) {
          // Handle error when checking increase eligibility
          console.error("Error checking increase eligibility:", error);
          Alert.alert("Success", "Exercise completed successfully", [
            { text: "OK", onPress: () => navigation.goBack() },
          ]);
        }
      } else if (painLevel > 3) {
        // If pain level is high, ask if they want to decrease difficulty
        Alert.alert(
          "Exercise Difficulty",
          "Your pain level is high. Would you like to decrease the difficulty for next time?",
          [
            {
              text: "No",
              onPress: () => navigation.goBack(),
              style: "cancel",
            },
            {
              text: "Yes",
              onPress: async () => {
                try {
                  await api.post(
                    `/user-exercises/${userExerciseId}/confirm_decrease/`,
                    { confirm: "yes" }
                  );
                  Alert.alert(
                    "Success",
                    "Exercise difficulty decreased for next time!"
                  );
                  navigation.goBack();
                } catch (error) {
                  console.error("Error decreasing difficulty:", error);
                  Alert.alert(
                    "Error",
                    "Failed to decrease difficulty. Please try again."
                  );
                  navigation.goBack();
                }
              },
            },
          ]
        );
      } else {
        Alert.alert("Success", "Exercise completed successfully", [
          { text: "OK", onPress: () => navigation.goBack() },
        ]);
      }
    } catch (error) {
      // Handle error when saving exercise completion
      console.error("Error saving exercise:", error);
      Alert.alert(
        "Error",
        "Failed to save exercise completion. Please try again."
      );
    }
  };

  // Generate dropdown data
  const getDropdownData = (maxValue: number) => {
    return Array.from({ length: maxValue }, (_, i) => ({
      label: `${i + 1}`,
      value: i + 1,
    }));
  };

  // Generate pain level dropdown data
  const getPainLevelData = () => {
    return Array.from({ length: 11 }, (_, i) => ({
      label: `${i}`,
      value: i,
    }));
  };

  return {
    showCompletionForm,
    setShowCompletionForm,
    reps,
    setReps,
    sets,
    setSets,
    painLevel,
    setPainLevel,
    handleSave,
    getDropdownData,
    getPainLevelData,
  };
}
