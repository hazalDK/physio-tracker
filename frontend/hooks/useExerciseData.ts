import { useState, useEffect } from "react";
import { Alert } from "react-native";
import axios from "axios";
import { ExerciseItem, UserExerciseItem } from "@/types/exercise";
import { useAuth } from "./useAuth";

export function useExerciseData(
  exerciseId: number,
  userExerciseId: number | null
) {
  const [loading, setLoading] = useState(true);
  const [exercise, setExercise] = useState<ExerciseItem>();
  const [userExercise, setUserExercise] = useState<UserExerciseItem>();
  const { createApiInstance, refreshToken } = useAuth();

  useEffect(() => {
    const fetchData = async () => {
      setLoading(true);
      try {
        const api = await createApiInstance();
        if (!api) return;

        try {
          const [userExercisesResponse, exercisesResponse] = await Promise.all([
            api.get(`/user-exercises/${userExerciseId}/`),
            api.get(`/exercises/${exerciseId}/`),
          ]);

          setUserExercise(
            !Array.isArray(userExercisesResponse.data)
              ? userExercisesResponse.data
              : ({} as UserExerciseItem)
          );

          setExercise(
            !Array.isArray(exercisesResponse.data)
              ? exercisesResponse.data
              : ({} as ExerciseItem)
          );
        } catch (error) {
          if (axios.isAxiosError(error) && error.response?.status === 401) {
            const newToken = await refreshToken();
            if (!newToken) return;

            const api = await createApiInstance();
            if (!api) return;

            const [retryUserExercises, retryExercises] = await Promise.all([
              api.get(`/user-exercises/${userExerciseId}/`),
              api.get(`/exercises/${exerciseId}/`),
            ]);

            setUserExercise(
              !Array.isArray(retryUserExercises.data)
                ? retryUserExercises.data
                : ({} as UserExerciseItem)
            );

            setExercise(
              !Array.isArray(retryExercises.data)
                ? retryExercises.data
                : ({} as ExerciseItem)
            );
          } else {
            console.error("API request failed:", error);
            Alert.alert(
              "Error",
              "Failed to load your exercises. Please check your connection and try again."
            );
            setExercise({} as ExerciseItem);
            setUserExercise({} as UserExerciseItem);
          }
        }
      } catch (error) {
        console.error("Failed to load exercise data:", error);
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [exerciseId, userExerciseId]);

  return { loading, exercise, userExercise };
}
