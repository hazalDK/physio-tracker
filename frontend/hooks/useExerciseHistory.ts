import { useState, useEffect } from "react";
import axios from "axios";
import { useAuth } from "./useAuth";
import { ExerciseHistoryItem } from "@/types/report";

export function useExerciseHistory() {
  const [loading, setLoading] = useState(true);
  const [history, setHistory] = useState<ExerciseHistoryItem[]>([]);
  const [error, setError] = useState<string | null>(null);

  const { createApiInstance, refreshToken } = useAuth();

  const fetchHistory = async () => {
    setLoading(true);
    try {
      const api = await createApiInstance();
      if (!api) return;

      try {
        // Fetch exercise history from the API
        const response = await api.get("/reports/exercise_history/");
        if (response.data && response.data.history) {
          setHistory(response.data.history);
        }
      } catch (error) {
        // Handle 401 error by refreshing the token and retrying the request
        if (axios.isAxiosError(error) && error.response?.status === 401) {
          // Attempt token refresh
          const newToken = await refreshToken();
          if (!newToken) return;

          const api = await createApiInstance();
          if (!api) return;

          // Retry with new token
          const response = await api.get("/reports/exercise_history/");
          if (response.data && response.data.history) {
            setHistory(response.data.history);
          }
        } else {
          // Handle other errors
          console.error("API request failed:", error);
          setError(
            "Failed to load your exercise history. Please check your connection and try again."
          );
        }
      }
    } catch (error) {
      console.error("Failed to load exercise history.");
      setError("An error occurred while loading exercise history.");
    } finally {
      setLoading(false);
    }
  };

  // Fetch history on component mount
  useEffect(() => {
    fetchHistory();
  }, []);

  return {
    loading,
    history,
    error,
    refreshHistory: fetchHistory,
  };
}
