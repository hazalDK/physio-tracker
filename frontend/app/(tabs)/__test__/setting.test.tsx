import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import { Alert } from "react-native";
import Settings from "../settings";
import { useNavigation } from "@react-navigation/native";
import { useAuth } from "@/hooks/useAuth";
import * as SecureStore from "expo-secure-store";
import axios from "axios";

// Mock the hooks and modules
jest.mock("@react-navigation/native");
jest.mock("@/hooks/useAuth");
jest.mock("expo-secure-store");
jest.mock("axios");
jest.mock("../../../components/ReminderComponent", () => "ReminderComponent");

// Mock Alert
jest.spyOn(Alert, "alert").mockImplementation(() => {});

describe("Settings Component", () => {
  // Setup mock data and functions
  const mockNavigation = {
    navigate: jest.fn(),
  };

  const mockApi = {
    put: jest.fn(),
  };

  const mockCreateApiInstance = jest.fn().mockResolvedValue(mockApi);

  beforeEach(() => {
    // Reset mocks before each test
    jest.clearAllMocks();

    // Setup default mocks
    (useNavigation as jest.Mock).mockReturnValue(mockNavigation);
    (useAuth as jest.Mock).mockReturnValue({
      createApiInstance: mockCreateApiInstance,
    });
  });

  it("renders component correctly", () => {
    const { getByText } = render(<Settings />);

    expect(getByText("Customisable Reminder:")).toBeTruthy();
    expect(getByText("Privacy")).toBeTruthy();
    expect(getByText("Update password")).toBeTruthy();
    expect(getByText("Sign Out")).toBeTruthy();
  });

  it("opens password modal when update password button is pressed", () => {
    const { getByText, queryByText } = render(<Settings />);

    // Modal should not be visible initially
    expect(queryByText("Update Password")).toBeNull();

    // Press the button to open modal
    fireEvent.press(getByText("Update password"));

    // Modal should now be visible
    expect(getByText("Update Password")).toBeTruthy();
    expect(getByText("Current Password:")).toBeTruthy();
    expect(getByText("New Password:")).toBeTruthy();
    expect(getByText("Confirm New Password:")).toBeTruthy();
  });

  it("handles sign out correctly", async () => {
    // Mock SecureStore.deleteItemAsync to resolve successfully
    (SecureStore.deleteItemAsync as jest.Mock)
      .mockResolvedValueOnce(undefined) // for access_token
      .mockResolvedValueOnce(undefined); // for refresh_token

    const { getByText } = render(<Settings />);

    fireEvent.press(getByText("Sign Out"));

    await waitFor(() => {
      expect(SecureStore.deleteItemAsync).toHaveBeenCalledWith("access_token");
      expect(SecureStore.deleteItemAsync).toHaveBeenCalledWith("refresh_token");
      expect(mockNavigation.navigate).toHaveBeenCalledWith("login");
      expect(Alert.alert).toHaveBeenCalledWith(
        "Signed out",
        "You have been successfully signed out."
      );
    });
  });

  it("handles sign out error correctly", async () => {
    // Mock SecureStore.deleteItemAsync to reject with an error
    (SecureStore.deleteItemAsync as jest.Mock).mockRejectedValue(
      new Error("Failed to delete")
    );

    const { getByText } = render(<Settings />);

    fireEvent.press(getByText("Sign Out"));

    await waitFor(() => {
      expect(Alert.alert).toHaveBeenCalledWith(
        "Error",
        "Failed to sign out. Please try again."
      );
    });
  });

  it("toggles password visibility when show/hide button is pressed", () => {
    const { getByText, getAllByPlaceholderText } = render(<Settings />);

    // First open the modal
    fireEvent.press(getByText("Update password"));

    // All password fields should start with secureTextEntry=true
    const passwordFields = getAllByPlaceholderText(/password/i);
    passwordFields.forEach((field) => {
      expect(field.props.secureTextEntry).toBe(true);
    });

    // Toggle password visibility
    fireEvent.press(getByText("👁️ Show Passwords"));

    // All password fields should now have secureTextEntry=false
    passwordFields.forEach((field) => {
      expect(field.props.secureTextEntry).toBe(false);
    });

    // Toggle back
    fireEvent.press(getByText("🙈 Hide Passwords"));

    // All password fields should be secure again
    passwordFields.forEach((field) => {
      expect(field.props.secureTextEntry).toBe(true);
    });
  });

  it("validates all fields are required when updating password", async () => {
    const { getByText } = render(<Settings />);

    // Open the modal
    fireEvent.press(getByText("Update password"));

    // Try to update without filling in any fields
    fireEvent.press(getByText("Update"));

    expect(Alert.alert).toHaveBeenCalledWith(
      "Error",
      "All fields are required."
    );
    expect(mockCreateApiInstance).not.toHaveBeenCalled();
  });

  it("validates passwords match when updating password", async () => {
    const { getByText, getByPlaceholderText } = render(<Settings />);

    // Open the modal
    fireEvent.press(getByText("Update password"));

    // Fill in fields but with mismatched new passwords
    fireEvent.changeText(
      getByPlaceholderText("Enter current password"),
      "oldpassword"
    );
    fireEvent.changeText(
      getByPlaceholderText("Enter new password"),
      "Newpassword1!"
    );
    fireEvent.changeText(
      getByPlaceholderText("Confirm new password"),
      "DifferentPassword1!"
    );

    // Try to update
    fireEvent.press(getByText("Update"));

    expect(Alert.alert).toHaveBeenCalledWith(
      "Error",
      "New passwords do not match."
    );
    expect(mockCreateApiInstance).not.toHaveBeenCalled();
  });

  it("validates new password differs from current password", async () => {
    const { getByText, getByPlaceholderText } = render(<Settings />);

    // Open the modal
    fireEvent.press(getByText("Update password"));

    // Fill in fields with same current and new password
    fireEvent.changeText(
      getByPlaceholderText("Enter current password"),
      "Password123!"
    );
    fireEvent.changeText(
      getByPlaceholderText("Enter new password"),
      "Password123!"
    );
    fireEvent.changeText(
      getByPlaceholderText("Confirm new password"),
      "Password123!"
    );

    // Try to update
    fireEvent.press(getByText("Update"));

    expect(Alert.alert).toHaveBeenCalledWith(
      "Error",
      "New password cannot be the same as current password."
    );
    expect(mockCreateApiInstance).not.toHaveBeenCalled();
  });

  it("validates password complexity requirements", async () => {
    const { getByText, getByPlaceholderText } = render(<Settings />);

    // Open the modal
    fireEvent.press(getByText("Update password"));

    // Fill in fields with weak password
    fireEvent.changeText(
      getByPlaceholderText("Enter current password"),
      "oldpassword"
    );
    fireEvent.changeText(getByPlaceholderText("Enter new password"), "simple");
    fireEvent.changeText(
      getByPlaceholderText("Confirm new password"),
      "simple"
    );

    // Try to update
    fireEvent.press(getByText("Update"));

    expect(Alert.alert).toHaveBeenCalledWith(
      "Error",
      "Password must contain:\n- 8+ characters\n- 1 uppercase\n- 1 lowercase\n- 1 number\n- 1 special character"
    );
    expect(mockCreateApiInstance).not.toHaveBeenCalled();
  });

  it("successfully updates password", async () => {
    // Mock successful API response
    mockApi.put.mockResolvedValue({ status: 200 });

    const { getByText, getByPlaceholderText } = render(<Settings />);

    // Open the modal
    fireEvent.press(getByText("Update password"));

    // Fill in fields with valid data
    fireEvent.changeText(
      getByPlaceholderText("Enter current password"),
      "OldPassword123!"
    );
    fireEvent.changeText(
      getByPlaceholderText("Enter new password"),
      "NewPassword456@"
    );
    fireEvent.changeText(
      getByPlaceholderText("Confirm new password"),
      "NewPassword456@"
    );

    // Update password
    fireEvent.press(getByText("Update"));

    await waitFor(() => {
      expect(mockCreateApiInstance).toHaveBeenCalled();
      expect(mockApi.put).toHaveBeenCalledWith(
        "http://192.168.68.111:8000/users/update_password/",
        {
          current_password: "OldPassword123!",
          new_password: "NewPassword456@",
        }
      );
      expect(Alert.alert).toHaveBeenCalledWith(
        "Success",
        "Password updated successfully!"
      );
    });
  });

  it("handles incorrect current password error", async () => {
    // Mock API error response for incorrect password
    const error = {
      response: {
        status: 400,
        data: { error: "Current password is incorrect" },
      },
      isAxiosError: true,
    };
    mockApi.put.mockRejectedValue(error);
    axios.isAxiosError.mockReturnValue(true);

    const { getByText, getByPlaceholderText } = render(<Settings />);

    // Open the modal
    fireEvent.press(getByText("Update password"));

    // Fill in fields
    fireEvent.changeText(
      getByPlaceholderText("Enter current password"),
      "WrongPassword123!"
    );
    fireEvent.changeText(
      getByPlaceholderText("Enter new password"),
      "NewPassword456@"
    );
    fireEvent.changeText(
      getByPlaceholderText("Confirm new password"),
      "NewPassword456@"
    );

    // Try to update
    fireEvent.press(getByText("Update"));

    await waitFor(() => {
      expect(mockCreateApiInstance).toHaveBeenCalled();
      expect(mockApi.put).toHaveBeenCalled();
      expect(Alert.alert).toHaveBeenCalledWith(
        "Incorrect Password",
        "The password you entered doesn't match your current password. Please try again."
      );
    });
  });

  it("handles generic API error", async () => {
    // Mock API error response
    const error = {
      response: {
        status: 500,
        data: { error: "Server error" },
      },
      isAxiosError: true,
    };
    mockApi.put.mockRejectedValue(error);
    axios.isAxiosError.mockReturnValue(true);

    const { getByText, getByPlaceholderText } = render(<Settings />);

    // Open the modal
    fireEvent.press(getByText("Update password"));

    // Fill in fields
    fireEvent.changeText(
      getByPlaceholderText("Enter current password"),
      "OldPassword123!"
    );
    fireEvent.changeText(
      getByPlaceholderText("Enter new password"),
      "NewPassword456@"
    );
    fireEvent.changeText(
      getByPlaceholderText("Confirm new password"),
      "NewPassword456@"
    );

    // Try to update
    fireEvent.press(getByText("Update"));

    await waitFor(() => {
      expect(mockCreateApiInstance).toHaveBeenCalled();
      expect(mockApi.put).toHaveBeenCalled();
      expect(Alert.alert).toHaveBeenCalledWith("Error", "Server error");
    });
  });

  it("handles network error", async () => {
    // Mock network error
    mockApi.put.mockRejectedValue(new Error("Network Error"));
    axios.isAxiosError.mockReturnValue(false);

    const { getByText, getByPlaceholderText } = render(<Settings />);

    // Open the modal
    fireEvent.press(getByText("Update password"));

    // Fill in fields
    fireEvent.changeText(
      getByPlaceholderText("Enter current password"),
      "OldPassword123!"
    );
    fireEvent.changeText(
      getByPlaceholderText("Enter new password"),
      "NewPassword456@"
    );
    fireEvent.changeText(
      getByPlaceholderText("Confirm new password"),
      "NewPassword456@"
    );

    // Try to update
    fireEvent.press(getByText("Update"));

    await waitFor(() => {
      expect(mockCreateApiInstance).toHaveBeenCalled();
      expect(mockApi.put).toHaveBeenCalled();
      expect(Alert.alert).toHaveBeenCalledWith(
        "Error",
        "Network error. Please check your connection and try again."
      );
    });
  });

  it("disables inputs and buttons when updating is in progress", async () => {
    // Use a promise that we control to keep the updating state active
    let resolveApiCall: (arg0: { status: number }) => void;
    const apiPromise = new Promise((resolve) => {
      resolveApiCall = resolve;
    });
    mockApi.put.mockReturnValue(apiPromise);

    const { getByText, getByPlaceholderText } = render(<Settings />);

    // Open the modal
    fireEvent.press(getByText("Update password"));

    // Fill in fields
    fireEvent.changeText(
      getByPlaceholderText("Enter current password"),
      "OldPassword123!"
    );
    fireEvent.changeText(
      getByPlaceholderText("Enter new password"),
      "NewPassword456@"
    );
    fireEvent.changeText(
      getByPlaceholderText("Confirm new password"),
      "NewPassword456@"
    );

    // Start update (but don't resolve yet)
    fireEvent.press(getByText("Update"));

    // While updating, elements should be disabled
    await waitFor(() => {
      expect(getByText("Updating...")).toBeTruthy();
      expect(
        getByPlaceholderText("Enter current password").props.editable
      ).toBe(false);
      expect(getByPlaceholderText("Enter new password").props.editable).toBe(
        false
      );
      expect(getByPlaceholderText("Confirm new password").props.editable).toBe(
        false
      );
      expect(getByText("Updating...")).toBeTruthy();

      // Now resolve the API call
      resolveApiCall({ status: 200 });
    });
  });
});
