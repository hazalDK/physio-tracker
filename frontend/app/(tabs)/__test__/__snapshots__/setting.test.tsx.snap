// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings Component renders correctly with data 1`] = `null`;
